### PR TITLE
Added fix and test for bug #3710 and fix to renderer.nim

### PR DESCRIPTION
**Pull request by skyfex**
This fixes bug #3710.

All tests pass in my build.

But I'm not super confident about the fix, so someone more familiar with the compiler should review it.

The fix in renderer.nim should provide clearer error messages for similar errors in the future. Previously the code `type(x).name`, i.e. the AST `nkDotExpr(nkTypeOfExpr("x"), "name")` would be rendered as `type x.name`. This fix adds parens so it renders back as `type(x).name`. It seems to me that this is more unambiguous and more consistent with prevelant coding style. If this fix is not desired I can create a new pull request without it.

### CODE REVIEW

File: compiler/semexprs.nim [hunk 1]
```diff
@@ -1060,7 +1060,7 @@ proc builtinFieldAccess(c: PContext, n: PNode, flags: TExprFlags): PNode =
     of tyTypeParamsHolders:
       return readTypeParameter(c, ty, i, n.info)
     of tyObject, tyTuple:
-      if ty.n != nil and ty.n.kind == nkRecList:
+      if (not isTypeExpr(n.sons[0])) and ty.n != nil and ty.n.kind == nkRecList:
```
Comment (Araq) [inline compiler/semexprs.nim]: Why is that necessary?

Comment (skyfex) [inline compiler/semexprs.nim]: All I really know is that it fixes the issue and it doesn't seem to break
any tests. I don't understand the compiler well enough to reason about the
code globally.

But as far as I can tell the code is not necessary to implement the
semantics of type(x.name) (as seems to be the purpose), but messes up the
interpretation of type(x).name if x is an object containing the field name.

But it could be totally wrong.

It would be nice if there were more comments in the compiler. It's a bit
hard for an outsider to get into.

**Comment by Araq** [comment]
The fix is not correct and I have merged the renderer improvements manually.
